Racing

teccheck opened the pull request:
Left to do
* [x] Boost behavior
* [x] Ranking updates throughout race (existing checkpoint detection system can be used for this)
* [x] End of the race:
  * [x] Display the correct times & ranking in leaderboard
  * [x] Show ranking after race
  * [x] Add handlers for the buttons there
  * [x] Give rewards
  * [x] Close the zone instance
* [x] Return to correct zone + location for each racing world (currently always goes to NS)
* [x] Add check to kill cars that have fallen off map (which can happen for example in the FV one)
* [x] Investigate into random kill bug
* [x] Correctly handle players trying to join after race has started
* [x] Progress missions and achievements
  * [x] Fix Velocity Lane (you can claim rewards without finishing missions)
  * [x] Find out why some achievements don't work
  * [x] Call `RacingSmashAsync`
  * [x] Call `RacingMissionCompleteAsync`
* [x] Detect player going the wrong way and reset their position
* [x] Respawning
  * [x] Make it work
  * [x] Add small delay (check old footage for this)
* [x] Imagination orb sync
* [x] Configure Server to not accept any more players after race started
* [x] Wait for players in the minigame queue (optional?)
  * [x] Communicate players from queue to RacingControlComponent
* [ ] (Optional) Gameplay config for minPlayerCount
* [ ] What are these [lap colliders](https://explorer.lu-dev.net/objects/9534/5)?
  * [ ] Implement them? 
* [x] Handle players leaving

Scripts:
* [x] All: Imagination crates
* [x] FV: Imagination Dragon Eggs
  * [x] Give imagination
  * [x] Some sort of respawning management script
* [x] FV: Geyser things in the refinery
* [x] FV: Moving platforms (changing paths each lap)

Misc:
* [x] Description for `_arrayNoLenght`
* [ ] Revert testing values
  * [x] LapCount (3)
  * [ ] MinPlayerCount in RacingTask (3)

enteryournamehere committed:
Add component structs

enteryournamehere committed:
Some experimenting

enteryournamehere committed:
Fix Instantiate when item is not in Objects table

enteryournamehere committed:
Add OnAcknowledgePossession event to GameObject

teccheck committed:
Merge branch 'dev' into enhancement/racing

teccheck committed:
Add script for mission 'Build Your Ride!'

teccheck committed:
Remove logs from NimbusCarBuildMission.cs

Jettford commented:
I ask that you please credit Simon in this code (apologies if I missed you already doing so), he has already has his racing code reimplemented without credit

Jettford commented:
and I missed it..... thanks for leaving him credit, it was a lot of work that he put into racing.

teccheck committed:
Fix LDF parsing in game messages

teccheck committed:
Fix for (parts of) racing control component

teccheck committed:
Use players car for racing

enteryournamehere committed:
One small step for car(?)

This probably includes some unnecessary changes

teccheck committed:
Implement racing start routine

teccheck committed:
Revert testing changees

enteryournamehere committed:
Merge two lists and change some names

enteryournamehere committed:
Remove position update logging

enteryournamehere committed:
Merge branch 'dev' into enhancement/racing

enteryournamehere committed:
Imagination spawners

teccheck committed:
Implement exit button for racing

enteryournamehere committed:
Fix ModuleAssembly serialization (fix imagination)

enteryournamehere committed:
Position updates

enteryournamehere committed:
Respawning

teccheck committed:
Implement CarBoostBehavior

enteryournamehere committed:
Fix FindItem for first item in inventory

enteryournamehere committed:
Ignore vehicle object in /near

enteryournamehere committed:
Reset car if it has fallen off the map

teccheck committed:
Add zone return data for racing

enteryournamehere committed:
Communicate car boost to others

teccheck committed:
Fix position and roatation before race

enteryournamehere committed:
More boost communication

enteryournamehere committed:
Remove position log

enteryournamehere committed:
Add vehicle death trigger script

enteryournamehere committed:
Initialize FV racetrack things to correct position

enteryournamehere committed:
Imagination crate script

enteryournamehere committed:
FV racetrack dragon eggs imagination script

enteryournamehere committed:
Remove array to fix ranking, thanks @TecCheck

teccheck committed:
Add delay to respawn in racing

enteryournamehere committed:
End of race exit button

enteryournamehere committed:
Better timing at start of race

enteryournamehere committed:
Avoid null reference in imagination handler

teccheck committed:
Fix match queue display

enteryournamehere committed:
Send players to match instance when zone is loaded

enteryournamehere committed:
Actually start timer

enteryournamehere committed:
Remove some unnecessary usages of .ToArray()

enteryournamehere committed:
Disable logging every message sent/received

teccheck committed:
Add RacingTask

teccheck committed:
Implement timing and ranking for racing

teccheck committed:
Fix RacingTask

teccheck committed:
Implement more stuff in lap routine

teccheck committed:
Add racing leaderboard

teccheck committed:
Add migrations for racing leaderboards

teccheck committed:
Implement rewards for racing

teccheck committed:
Fix BestLapTime being 0

teccheck committed:
Close race instance after all players have left

TheNexusAvenger commented on Uchu.Core/Extensions/RakConnectionExtensions.cs:
```diff
@@ -16,7 +16,7 @@ public static void Send(this IRakConnection @this, ISerializable serializable)
                 throw new ArgumentNullException(nameof(serializable), 
                     ResourceStrings.RakConnectionExtensions_Send_StreamNullException);
-            Logger.Debug($"Sending {serializable}");
+            // Logger.Debug($"Sending {serializable}");
```
Is there a reason why several debug statements were taken out? This applies to several files but I am only putting 1 comment for it.

enteryournamehere replied:
For racing it's constantly serializing people's positions which makes the logs kind of unreadable

TheNexusAvenger replied:
I see. In that case, I see 2 options: we remove the logging statements altogether, or we start using the `Trace` log level.

TheNexusAvenger commented on Uchu.Core/Serializable/Struct/Property/PacketProperty.cs:
```diff
@@ -88,6 +92,8 @@ public class PacketProperty : IPacketProperty
         /// </summary>
         private Type _arrayLengthPropertyType = typeof(uint);
 
+        private bool _arrayNoLenght = false;
```
Can a docstring be added to this to be consistent with the surrounding attributes?

teccheck replied:
Sure. I'll add that to the TODO list

TheNexusAvenger commented on Uchu.World/Objects/Components/ReplicaComponents/RacingControlComponent.cs:
```diff
@@ -1,21 +1,713 @@
+            return packet;
+        }
+
+        private enum RacingStatus {
```
I assume these classes will be moved out like the other packets at some point?

teccheck replied:
I'm not sure about this one. This is supposed to be an internal thing. It has no use outside of RacingControlComponent iirc

TheNexusAvenger commented on Uchu.World/Objects/Components/ReplicaComponents/VehiclePhysicsComponent.cs:
```diff
@@ -1,24 +1,121 @@
-}
\ No newline at end of file
+
+    public struct VehiclePhysicsConstruction {
```
I assume these classes will be moved out like the other packets at some point?

teccheck replied:
I think that would be good

TheNexusAvenger commented on Uchu.World/Objects/Components/ReplicaComponents/VehiclePhysicsComponent.cs:
```diff
@@ -1,24 +1,121 @@
+    public struct VehiclePhysicsSerialization {
+        [Default]
+        public VehicleFrameStatsTeleportInfo VehicleFrameStatsTeleportInfo { get; set; }
+        // [Default] bool with flag? should this be here
```
`bool`s having the `[Default]` attribute is supported, because "LU moment".

TheNexusAvenger commented on Uchu.World/Packets/Replica/Serialize/ModuleAssemblySerialization.cs:
```diff
@@ -0,0 +1,26 @@
+    {
+        public bool Flag { get; set; }
+        [Requires("Flag")]
+        public ModuleAssemblyInfo ModuleAssemblyInfo { get; set; }
```
Any reason the flag is specified and required instead of just:
```C#
[Default]
public ModuleAssemblyInfo ModuleAssemblyInfo { get; set; }
```

enteryournamehere replied:
That seems to be broken for struct properties as are used here, see https://discord.com/channels/640928584113979402/640928584113979404/909078632268443660

TheNexusAvenger reviewed (commented):
Really like where this is going. While this isn't done, I am posting a partial review of the current code now.

teccheck committed:
Add description to _arrayNoLenght attribute

teccheck committed:
Implement PostRacePlayerInfos

teccheck committed:
Format RacingControlComponent

teccheck committed:
Refactor RacingTask

enteryournamehere committed:
Merge branch 'dev' of github.com:UchuServer/Uchu into enhancement/racing

teccheck committed:
Implement last racing task types

Frozenreflex committed:
Script changes

Frozenreflex committed:
Add progression for "Dragonscourge" and "Terror of Dragonmaw Chasm"

Frozenreflex committed:
Add blue mark and add race_imagine_powerup to script blacklist
remove the scripted smash task because I realized this should be done already

Frozenreflex committed:
Add maelstrom geiser script

Frozenreflex committed:
Add dragon animation and dragon egg manager scripts

teccheck committed:
Clean up RacingControlComponent

teccheck committed:
Reset player when driving wrong way

Frozenreflex committed:
Oops

teccheck committed:
Merge branch 'dev' into enhancement/racing

teccheck committed:
Don't accept any new players when race has started

Frozenreflex committed:
Add pillar animations, add RigidBodyPhantomPhysicsComponent, fix dragon animations
Pillars don't collide yet, lap 3's dragon animation doesn't work yet

enteryournamehere committed:
Disable render distance filter for racing

Frozenreflex committed:
Fix and add to script blacklist, make death triggers in racing only kill once, add water vehicle death trigger
As far as I can tell, the water trigger is only used in the GF track, and I couldn't tell the difference, but it's there anyway.

teccheck committed:
Wait for players in minigame queue

teccheck committed:
Handle players leaving

teccheck committed:
Revert min player count in racing task to 3

TecCheck commented:
Can we merge this? I'd really like to finish this

enteryournamehere commented:
IIRC, there were some lingering issues with, uh, communicating position updates with multiple players, or loading of objects, or something, I think? It's been a good while. Either way, though, pretty much all of racing is implemented, so it's fine by me to merge as is :+1:

TecCheck commented:
I'd say we can improve racing later
